src/network/mcpe/handler/InGamePacketHandler.php

```diff
@@ -840,15 +842,33 @@ public function handlePlayerSkin(PlayerSkinPacket $packet) : bool{
 			$this->session->getLogger()->debug("Refused duplicate skin change request");
 			return true;
 		}
-		$this->lastRequestedFullSkinId = $packet->skin->getFullSkinId();
+		$currentFullSkinId = $this->lastRequestedFullSkinId = $packet->skin->getFullSkinId();
 
 		$this->session->getLogger()->debug("Processing skin change request");
-		try{
-			$skin = $this->session->getTypeConverter()->getSkinAdapter()->fromSkinData($packet->skin);
-		}catch(InvalidSkinException $e){
-			throw PacketHandlingException::wrap($e, "Invalid skin in PlayerSkinPacket");
-		}
-		return $this->player->changeSkin($skin, $packet->newSkinName, $packet->oldSkinName);
+		$this->player->getServer()->getAsyncPool()->submitTask(new ProcessSkinTask(
```

Comment:
This would probably make the server to freeze / crash if a malicious client floods tons of skin change requests per tick.

Reply:
Would the existing `$this->gamePacketLimiter` be sufficient for skin change packets, or would it be better to implement a separate rate limiter specifically for `PlayerSkinPacket` to prevent potential flooding?

Reply:
If a separate rate limiter is needed, how should we set the values for `INCOMING_SKIN_PACKET_PER_TICK = 1` and `INCOMING_SKIN_PACKET_BUFFER_TICKS = 100`?
Would these settings be appropriate, or should we adjust them further?

Reply:
> Would the existing `$this->gamePacketLimiter` be sufficient for skin change packets, or would it be better to implement a separate rate limiter specifically for `PlayerSkinPacket` to prevent potential flooding?

Probably not. It's supposed to limit all packets, so it has quite generous limits.

More so the issue here is memory overload if the client spams a lot of skin change packets (it could cause a large backlog in the worker pool & hog a lot of memory).

My thought was to limit the number of concurrent `ProcessSkinTask`s per session, but I'm not sure if this is the best way.

Reply:
Limiting the number of `PlayerSkinPacket` packets seems easier to implement and would have a similar effect to limiting the number of `ProcessSkinTasks`.

In my preliminary testing, setting `INCOMING_SKIN_PACKET_PER_TICK = 0.1` and `INCOMING_SKIN_PACKET_BUFFER_TICKS = 20` seems promising (though it still needs sufficient testing). However, this would require changing the `$averagePerTick` in `PacketLimiter` from an `int` to a `float` or both. Would that be acceptable?

Reply:
No, I don't think so. The number of packets isn't the only factor we need to consider. We also have to be aware of how long each packet takes to process.

Under normal conditions the packets might take 10 ms each to process, but an attacker could craft packets that take 10-100x longer to process and still be subject to the same packets-per-tick limits (getting more lag per packet). That's why I said we probably need to limit the number of tasks, rather than the number of packets.

Reply:
> No, I don't think so. The number of packets isn't the only factor we need to consider. We also have to be aware of how long each packet takes to process.
> 
> 
> 
> Under normal conditions the packets might take 10 ms each to process, but an attacker could craft packets that take 10-100x longer to process and still be subject to the same packets-per-tick limits (getting more lag per packet). That's why I said we probably need to limit the number of tasks, rather than the number of packets.

Aside from the better methods for now, we could consider handling these potential time-bomb data by spawning new processes, similar to ConsoleReaderChildProcess. This would allow the main thread to control their runtime, using OS-level process killing (with proper abstraction for generalization—I’ve already implemented this in my private project, leveraging opis/closure). However, this approach does come with challenging compatibility issues.

Reply:
That sounds overengineered to me.